### PR TITLE
More Viable Plasma Episode 1 — Youngest Inputs

### PR DESCRIPTION
**Pull request by troggy**
image

Relates to #21

### CODE REVIEW

File: contracts/ExitHandler.sol [hunk 1]
```diff
@@ -33,6 +35,7 @@ contract ExitHandler is DepositHandler {
     address owner;
     bool finalized;
     uint256 stake;
+    uint256 priorityTimestamp;
```
Comment (johannbarbie) [inline contracts/ExitHandler.sol]: make unt32 and move next to bool, to use compact storage

File: contracts/ExitHandler.sol [hunk 2]
```diff
@@ -6,9 +6,11 @@
  * found in the LICENSE file in the root directory of this source tree.
  */
 
+/* solium-disable security/no-block-members */
+
 pragma solidity 0.4.24;
 
-import "openzeppelin-solidity/contracts/math/Math.sol";
+import "../node_modules/openzeppelin-solidity/contracts/math/Math.sol";
```
Comment (johannbarbie) [inline contracts/ExitHandler.sol]: why?

File: test/exitHandler.js [hunk 3]
```diff
@@ -17,6 +16,13 @@ const PriorityQueue = artifacts.require('PriorityQueue');
 const MintableToken = artifacts.require('MockMintableToken');
 const SpaceDustNFT = artifacts.require('SpaceDustNFT');
 
+const aSecond = async () => new Promise(resolve => setTimeout(resolve, 1000));
```
Comment (johannbarbie) [inline test/exitHandler.js]: 🎉

File: test/exitHandler.js [hunk 4]
```diff
@@ -85,196 +130,153 @@ contract('ExitHandler', (accounts) => {
...
-
+        const event = await exitHandler.startExit(inputProof, transferProof, 0, 0, { from: bob });
+        
+        const utxoId = exitUtxoId(event);
+        assert.equal(utxoId, spendTx.inputs[0].prevout.getUtxoId());
 
         // challenge exit and make sure exit is removed
-        let exit = await exitHandler.exits(outpoint.getUtxoId());
-        assert.equal(exit[2], bob);
+        assert.equal((await exitHandler.exits(utxoId))[2], bob);
```
Comment (johannbarbie) [inline test/exitHandler.js]: 🎉

File: test/exitHandler.js [hunk 5]
```diff
@@ -44,35 +104,20 @@ contract('ExitHandler', (accounts) => {
       // At this point alice is the owner of bridge and depositHandler and has 10000 tokens
       // Bob is the bridge operator and exitHandler and has 0 tokens
       // Note: all txs in these tests originate from alice unless otherwise specified
+
+      await seedTxs();
```
Comment (johannbarbie) [inline test/exitHandler.js]: 🎉

File: test/exitHandler.js [hunk 6]
```diff
@@ -44,35 +104,20 @@ contract('ExitHandler', (accounts) => {
...
+      await seedNftTxs();
```
Comment (johannbarbie) [inline test/exitHandler.js]: 🎉

File: test/fastExitHandler.js [hunk 7]
```diff
@@ -6,7 +6,9 @@
  * found in the LICENSE file in the root directory of this source tree.
  */
 
-import { Period, Block, Tx, Input, Output, Outpoint, Exit } from 'leap-core';
+import { Tx, Input, Output, Outpoint, Exit } from 'leap-core';
+
+import { submitNewPeriodWithTx } from './helpers';
```
Comment (johannbarbie) [inline test/fastExitHandler.js]: 🎉

File: contracts/ExitHandler.sol [hunk 8]
```diff
@@ -57,33 +60,47 @@ contract ExitHandler is DepositHandler {
...
+    // check youngest input tx inclusion in the root chain block
+    bytes32 inputTxHash;
+    (, inputTxHash,) = TxLib.validateProof(96, _youngestInputProof);
+    require(
+      inputTxHash == exitingTx.ins[_inputIndex].outpoint.hash, 
+      "Input from the proof is not referenced in exiting tx"
+    );
+    
     uint256 priority;
+    uint256 exitableAt = Math.max256(timestamp + (2 * exitDuration), block.timestamp + exitDuration);
```
Comment (johannbarbie) [inline contracts/ExitHandler.sol]: how about adding the position?
something like: `exitableAt = exitableAt << 32 | txPos`

**Comment by eezcjkr** [comment]
Please fix merge conflicts.

File: contracts/ExitHandler.sol (hunk 2 above)
Comment (troggy) [reply contracts/ExitHandler.sol]: to make code linter happy. Any problem with such a code?

Comment (johannbarbie) [reply contracts/ExitHandler.sol]: > @johannbarbie anything else here?
looks good

**Comment by johannbarbie** [comment]
> @johannbarbie anything else here?

looks great

**Pull request merged**